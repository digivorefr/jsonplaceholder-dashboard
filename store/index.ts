import { ActionTree, MutationTree, GetterTree } from 'vuex'

export type User = {
  id: number;
  name: string;
  username: string;
  email: string;
  address: {
    street: string;
    suite: string;
    city: string;
    zipcode: string;
    geo: {
      lat: string;
      lng: string;
    }
  };
  phone: string;
  website: string;
  company: {
    name: string;
    catchPhrase: string;
    bs: string;
  }
}

export type Status = 'initial' | 'pending' | 'loaded' | 'error';

export const state = () => ({
  status: 'initial' as Status,
  users: [] as User[]
})

export type UsersState = ReturnType<typeof state>;

export const mutations: MutationTree<UsersState> = {
  REFRESH: (state, newUsers: User[]) => (state.users = newUsers),
  UPDATE_STATUS: (state, newStatus: Status) => (state.status = newStatus)
}

export const getters: GetterTree<UsersState, UsersState> = {
  getUsers: (state: UsersState) => state.users,
  getUserById: (state: UsersState) => (id: number) => state.users.find(user => user.id === id)
}

export const actions: ActionTree<UsersState, UsersState> = {
  async fetchUsers({ commit }) {
    commit('UPDATE_STATUS', 'pending')
    try {
      const users = await this.$axios.$get('/users')
      commit('REFRESH', users)
      commit('UPDATE_STATUS', 'loaded')
    } catch {
      commit('UPDATE_STATUS', 'error')
    }
  }
}
