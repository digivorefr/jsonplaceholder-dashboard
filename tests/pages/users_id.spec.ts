import { shallowMount, Wrapper } from '@vue/test-utils'
import { User as UserState } from '@/store'
import User from '@/pages/users/_id.vue'

const routeMock = { params: { id: 1 } }
const usersMock = <UserState[]>[
  {
    id: 1,
    name: 'Leanne Graham',
    username: 'Bret',
    email: 'Sincere@april.biz',
    address: {
      street: 'Kulas Light',
      suite: 'Apt. 556',
      city: 'Gwenborough',
      zipcode: '92998-3874',
      geo: {
        lat: '-37.3159',
        lng: '81.1496'
      }
    },
    phone: '1-770-736-8031 x56442',
    website: 'hildegard.org',
    company: {
      name: 'Romaguera-Crona',
      catchPhrase: 'Multi-layered client-server neural-net',
      bs: 'harness real-time e-markets'
    }
  },
  {
    id: 2,
    name: 'Ervin Howell',
    username: 'Antonette',
    email: 'Shanna@melissa.tv',
    address: {
      street: 'Victor Plains',
      suite: 'Suite 879',
      city: 'Wisokyburgh',
      zipcode: '90566-7771',
      geo: {
        lat: '-43.9509',
        lng: '-34.4618'
      }
    },
    phone: '010-692-6593 x09125',
    website: 'https://www.anastasia.net',
    company: {
      name: 'Deckow-Crist',
      catchPhrase: 'Proactive didactic contingency',
      bs: 'synergize scalable supply-chains'
    }
  }
]
const storeMock = {
  state: {
    users: usersMock
  },
  dispatch: jest.fn(),
  getters: {
    getUserById: (): UserState | undefined => usersMock[0]
  }
}

const nuxtErrorMock = jest.fn()

describe('User', () => {
  test('renders correctly', async () => {
    const wrapper = shallowMount(User, {
      mocks: {
        $route: routeMock,
        $store: storeMock,
        $nuxt: { error: jest.fn() }
      }
    })
    await wrapper.vm.$nextTick()
    expect(wrapper.vm).toBeTruthy()
    expect(wrapper.exists()).toBe(true)
    expect(wrapper.element).toMatchSnapshot()

    wrapper.destroy()
  })
  test('User does not exist, throw nuxt 404', async () => {
    const wrapper = shallowMount(User, {
      mocks: {
        $route: routeMock,
        $store: {
          ...storeMock,
          getters: {
            getUserById: () => undefined
          }
        },
        $nuxt: { error: nuxtErrorMock }
      }
    })
    await wrapper.vm.$nextTick()
    expect(nuxtErrorMock).toHaveBeenCalledWith({ statusCode: 404, message: 'User not found' })
    wrapper.destroy()
  })
  test('Url cleaner', async () => {
    const wrapper: Wrapper<Vue & { [key: string]: any }> = shallowMount(User, {
      mocks: {
        $route: routeMock,
        $store: storeMock,
        $nuxt: { error: jest.fn() }
      }
    })
    await wrapper.vm.$nextTick()

    expect(wrapper.vm.cleanUrl('www.cafe80.com')).toBe('https://www.cafe80.com')
    expect(wrapper.vm.cleanUrl('http://biffcompany.com')).toBe('http://biffcompany.com')
  })
  test('phone number cleaner', async () => {
    const wrapper: Wrapper<Vue & { [key: string]: any }> = shallowMount(User, {
      mocks: {
        $route: routeMock,
        $store: storeMock,
        $nuxt: { error: jest.fn() }
      }
    })
    await wrapper.vm.$nextTick()

    expect(wrapper.vm.cleanPhone('3324564569')).toBe('+13324564569')
    expect(wrapper.vm.cleanPhone('1 332-456-4569')).toBe('+13324564569')
    expect(wrapper.vm.cleanPhone('+33 33 24 56 45 69')).toBe('+333324564569')
    expect(wrapper.vm.cleanPhone('(332).456.4569')).toBe('+13324564569')
    expect(wrapper.vm.cleanPhone('(332).456.4569  x4598')).toBe('+13324564569')
    expect(wrapper.vm.cleanPhone('123456')).toBe('')
    expect(wrapper.vm.cleanPhone('aghkjhga')).toBe('')
  })
})
