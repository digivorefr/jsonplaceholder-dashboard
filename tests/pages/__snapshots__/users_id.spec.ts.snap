// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User renders correctly 1`] = `
<div
  class="container container--grid"
>
  <div
    class="user"
  >
    <header>
      <div>
        <h1>
          Bret
        </h1>
         
        <p>
          #1 - Leanne Graham 
        </p>
      </div>
    </header>
     
    <section
      class="user__main"
    >
      <h2>
        Infos
      </h2>
       
      <div
        class="user__main__wrapper"
      >
        <div
          class="user__info"
        >
          <h4>
            Contact
          </h4>
           
          <ul>
            <li>
              <div
                class="icon icon--email"
              />
               
              <a
                href="mailto:Sincere@april.biz"
                target="_blank"
              >
                Sincere@april.biz
              </a>
            </li>
             
            <li>
              <div
                class="icon icon--phone"
              />
               
              <a
                href="tel:+17707368031"
                target="_blank"
              >
                1-770-736-8031 x56442
              </a>
            </li>
             
            <li>
              <div
                class="icon icon--link"
              />
               
              <a
                href="https://hildegard.org"
                target="_blank"
              >
                hildegard.org
              </a>
            </li>
          </ul>
        </div>
         
        <div
          class="user__info"
        >
          <h4>
            Address
          </h4>
           
          <ul>
            <li>
              <div
                class="icon icon--home"
              />
               
              <span>
                
                Apt. 556
                <br />
                
                Kulas Light
                <br />
                
                Gwenborough 92998
              
              </span>
            </li>
             
            <li>
              <div
                class="icon icon--location"
              />
               
              <a
                href="https://maps.google.com/?q=-37.3159,81.1496"
                target="_blank"
              >
                View on Google Maps
              </a>
            </li>
          </ul>
        </div>
         
        <div
          class="user__info"
        >
          <h4>
            Company
          </h4>
           
          <ul>
            <li>
              <div
                class="icon icon--business"
              />
               
              <span>
                <h5>
                  Romaguera-Crona
                </h5>
                 
                <span
                  class="catchphrase"
                >
                  
                  Multi-layered client-server neural-net
                
                </span>
              </span>
            </li>
             
            <li>
              <div
                class="icon icon--activity"
              />
              
              harness real-time e-markets
            
            </li>
          </ul>
        </div>
      </div>
    </section>
  </div>
   
  <div
    class="user__activities"
  >
    <posts-stub
      id="1"
    />
     
    <todos-stub
      id="1"
    />
     
    <albums-stub
      id="1"
    />
  </div>
</div>
`;
